common/src/main/java/com/microsoft/hydralab/performance/hprof/HeapProfProcessor.java

```diff
@@ -0,0 +1,449 @@
+    }
+
+    /**
+     * TODO
+     *
+     * @param instance
+     * @return
+     */
+    public static String getObjectRefTrace(Instance instance) {
+        return "todo";
+    }
+
+    public void registerExtractor(Extractor extractor) {
+        extractorMap.put(extractor.getName(), extractor);
+    }
+
+
+    public void loadAndExtract() throws IOException {
+        if (!heapDumpFile.exists()) {
+            return;
+        }
+
+        DataBuffer buffer = new MemoryMappedFileBuffer(heapDumpFile);
+
+        Snapshot snapshot = Snapshot.createSnapshot(buffer);
+        logger.info("start createSnapshot: " + heapDumpFile.getName());
+
+        deduplicateGcRoots(snapshot);
+        logger.info("deduplicateGcRoots: " + heapDumpFile.getName());
+
+        snapshot.computeDominators();
+        logger.info("compute dominators finished: " + heapDumpFile.getName());
+
+        List<Instance> dominatorList = snapshot.getReachableInstances();
+        if (dominatorList.isEmpty()) {
+            logger.warn("no gReachable Instances: in file: {}", heapDumpFile.getName());
+            return;
+        }
+        dominatorList.sort((instance1, instance2) -> -Long.compare(instance1.getTotalRetainedSize(), instance2.getTotalRetainedSize()));
+        logger.info("sortByRetainedSize finish start listing dominatorList: {}, in file: {}", dominatorList.size(), heapDumpFile.getName());
+
+        Iterator<Instance> instanceIterator = dominatorList.iterator();
+
+        int index = 0;
+        while (instanceIterator.hasNext()) {
+            Instance instance = instanceIterator.next();
+            if (instance.getClassObj() == null) {
+                continue;
+            }
+            processInstance(index++, instance);
+        }
+
+        // sort all extracted object info by retained size
+        for (Map.Entry<String, Extractor> entry : extractorMap.entrySet()) {
+            entry.getValue().onExtractComplete();
+        }
+    }
+
+    /**
+     * Pruning duplicates reduces memory pressure from hprof bloat added in Marshmallow.
+     */
+    private void deduplicateGcRoots(Snapshot snapshot) {
+        // THashMap has a smaller memory footprint than HashMap.
+        final THashMap<String, RootObj> uniqueRootMap = new THashMap<>();
+
+        final Collection<RootObj> gcRoots = snapshot.getGCRoots();
+        for (RootObj root : gcRoots) {
+            String key = generateRootKey(root);
+            if (!uniqueRootMap.containsKey(key)) {
+                uniqueRootMap.put(key, root);
+            }
+        }
+
+        // Repopulate snapshot with unique GC roots.
+        gcRoots.clear();
+        uniqueRootMap.forEach(new TObjectProcedure<String>() {
+            @Override
+            public boolean execute(String key) {
+                return gcRoots.add(uniqueRootMap.get(key));
+            }
+        });
+    }
+
+//    public File generateReport(File csvFile) {
```

Comment:
Can the reports be saved to the result folder? If not, I think we should make sure the `List` of `ObjectInfo` can be converted to JSON that contains enough information for Devs to locate the memory issue. e.g. Filed chains

Reply:
No, those code won't work. The ObjectInfo list has already passed to the `AndroidMemoryHprofResultParser`, it should be saved as part of parsedData.